### DISCUSSION TITLE
problem installing the windows community toolkit

### DISCUSSION THREAD

**Discussion by PalusiOld** [unoplatform/uno#13090] [Q&A] [answered]
i am using a uno platform project like you get when creating a blank uno project. I installed the "Uno.Microsoft.Toolkit.Uwp.UI.Controls.DataGrid" version of the toolkit to use a datagrid and folowed this guide:
https://platform.uno/docs/articles/uno-community-toolkit.html

My aim is to use the wrappanel for a crossplatform app. But I dont even can use the datagrid

My xaml code:
```xml
<Page x:Class="wrapgridtest.MainPage"
    xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
    xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
    xmlns:local="using:wrapgridtest"
    xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
    xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
    mc:Ignorable="d"
    xmlns:controls="using:CommunityToolkit.WinUI.UI.Controls"
    Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">

    <Grid>
        <controls:DataGrid>
             <controls:DataGrid.Columns>
                 <controls:DataGridTextColumn Header="Rank"/>
                 <controls:DataGridComboBoxColumn Header="Mountain"/>
             </controls:DataGrid.Columns>
         </controls:DataGrid>
    </Grid>
</Page>
```

When building the application I get the following errors:

`Error MSB4019: The imported project 'C:\Users\patri\.nuget\packages\uno.ui\4.0.0-dev.5688\buildTransitive\net6.0-android30.0\Uno.UI.SourceGenerators.props' was not found. Ensure that the expression in the import declaration ".\Uno.UI.SourceGenerators.props" is correct and that the file exists on disk.`

`Error MSB4019: The imported project 'C:\Users\patri\.nuget\packages\uno.ui\4.0.0-dev.5688\buildTransitive\net6.0-android30.0\Uno.UI.SourceGenerators.props' was not found. Ensure that the expression in the import declaration ".\Uno.UI.SourceGenerators.props" is correct and that the file exists on disk.`

`1>C:\Users\patri\source\repos\wrapgridtest\wrapgridtest\wrapgridtest\MainPage.xaml(13,14): XamlCompiler error WMC0001: Unknown type 'DataGridTextColumn' in XML namespace 'using:CommunityToolkit.WinUI.UI.Controls '`

I also cant install the "Uno.CommunityToolkit.WinUI" package using the vs nuget package manager.

Can someone help?

**Comment by PalusiOld** [comment]
I can now install the "Uno.CommunityToolkit.WinUI" package, but using this code i get the folowing errors:

```xml
<Page x:Class="DataGridTest.MainPage"
	xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
	xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
	xmlns:local="using:DataGridTest"
	xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
	xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
	mc:Ignorable="d"
    xmlns:controls="using:CommunityToolkit.WinUI.UI.Controls"
	Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">
    <Grid>
        <controls:DataGrid x:Name="dataGrid">
            <controls:DataGrid.Columns>
                <controls:DataGridTextColumn Header="Rank"/>
                <controls:DataGridComboBoxColumn Header="Mountain"/>
            </controls:DataGrid.Columns>
        </controls:DataGrid>
    </Grid>
</Page>
```

```
Severity Code Description Project File Line Suppression state
Error WMC9999 Type universe cannot resolve assembly: Uno.UI, Version=255.255.255.255, Culture=neutral, PublicKeyToken=null. DataGridTest C:\Users\patri\source\repos\DataGridTest\DataGridTest\DataGridTest\MainPage.xaml 1
```

```
Severity Code Description Project File Line Suppression state
Error XLS0503 A value of type DataGrid cannot be added to a collection or dictionary of type UIElementCollection. DataGridTest C:\Users\patri\source\repos\DataGridTest\DataGridTest\DataGridTest\MainPage.xaml 11
```

**Reply by jeromelaban** [reply]
Thanks for the report! Have you been able to build and run this particular sample: https://github.com/unoplatform/Uno.Samples/tree/master/UI/UnoWCTDataGridSample? This should give an actual example to start from.

That being said, it seems that the documentation may be out of date for you to get those errors.

**Comment by arnold256** [comment]
Was there any resolution to this I am having the exact same issue using CommunityToolkit and following the same documentation

**Reply by jeromelaban** [reply]
This will get fixed once the community toolkit 8.x is released, soon.

**Reply by tomchoondal** [reply]
@jeromelaban Any updates on the community toolkit throwing error while installing. We badly need community toolkit for our developments.

**Reply by jeromelaban** [reply]
@tomchoondal you can find updates here: https://github.com/CommunityToolkit/Windows/discussions/383. The latest previews are compatible with Uno Platform 5.x.

**Reply by agneszitte** [reply, marked as answer]
Hi @tomchoondal (cc @palusi, @arnold256) 
We will update soon [our documentation](https://platform.uno/docs/articles/uno-community-toolkit.html?tabs=tabid-winui) and samples with the new changes.

But here are the details with our [latest single project structure](https://platform.uno/blog/the-first-and-only-true-single-project-for-mobile-web-desktop-and-embedded-in-net/):

- For WCT (**v. 7.1.2 and below**):
	`xmlns:controls="using:CommunityToolkit.WinUI.UI.Controls"`
	
	In Directory.Packages.props:
	```
	<ItemGroup Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'windows'">
	  <PackageVersion Include="CommunityToolkit.WinUI.UI.Controls" Version="7.1.2" />
	  <!-- Add more community toolkit references here -->
	</ItemGroup>
	<ItemGroup Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) != 'windows'">
	  <PackageVersion Include="Uno.CommunityToolkit.WinUI.UI.Controls" Version="7.1.200" />
	  <!-- Add more uno community toolkit references here -->
	</ItemGroup>
	```
	
	In YourProjectName.csproj:
	```
	<ItemGroup Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) == 'windows'">
	  <PackageReference Include="CommunityToolkit.WinUI.UI.Controls" />
	  <!-- Add more community toolkit references here -->
	</ItemGroup>
	<ItemGroup Condition="$([MSBuild]::GetTargetPlatformIdentifier('$(TargetFramework)')) != 'windows'">
	  <PackageReference Include="Uno.CommunityToolkit.WinUI.UI.Controls" />
	  <!-- Add more uno community toolkit references here -->
	</ItemGroup>
	```
 
- For WCT (**v. 8.0 and above**):
  `xmlns:controls="using:CommunityToolkit.WinUI.Controls"`
    - **NOTE 1: The WCT 8.0 packages and above now include direct support for Uno Platform, simplifying the setup process.**
    - **NOTE 2: Updating to the latest WCT version requires an update to net8.0-windows10.0.22621 and Microsoft.WindowsAppSDK to the latest version (1.5.240311000).**
     (Related WCT issue: https://github.com/CommunityToolkit/Tooling-Windows-Submodule/issues/185)
	
	(Example using _CommunityToolkit.WinUI.Controls.Primitives_ here as I tested the WCT `WrapPanel`)
	In Directory.Packages.props:
	```
	 <ItemGroup>
	    <PackageVersion Include="CommunityToolkit.WinUI.Controls.Primitives" Version="8.1.240328-rc" />
	    <!-- Add more community toolkit references here -->
	 </ItemGroup>
	```
	
	In YourProjectName.csproj:
	```
	  <ItemGroup>
	    <PackageReference Include="CommunityToolkit.WinUI.Controls.Primitives" />
	    <!-- Add more community toolkit references here -->
	  </ItemGroup>
	```
 
I have made two sample apps to help the [community on discord](https://discord.com/channels/1182775715242967050/1182775715729526908/1237939787176869968) the other week for reference :
- Sample app using WCT (**v. 7.1.2**): [TestWrapPanelButtons.zip](https://github.com/unoplatform/uno/files/15425484/TestWrapPanelButtons.zip)
- Sample app using WCT (**v. 8.1.240328-rc**): [TestWrapPanelButtonsWCTRC.zip](https://github.com/unoplatform/uno/files/15425488/TestWrapPanelButtonsWCTRC.zip)

> [!IMPORTANT]
> **Here is the [Migration Guide from v7 to v8 for WCT](https://github.com/CommunityToolkit/Windows/wiki/Migration-Guide-from-v7-to-v8) for additional information on what changed lately between these versions.**
> As for example:
> - For the `DataGrid` control you will need to use **WCT version 7.1.2** as it is no longer a control available in the latest versions (v. 8.0 and above) of WCT. ([related details on Discord here](https://discord.com/channels/1182775715242967050/1193995186951426048/1221955960159993926))
> - For the `Carousel` control you will need to use **WCT version 7.1.2** as well as it is no longer a control available in the latest versions (v. 8.0 and above) of WCT. ([related details on Discord here](https://discord.com/channels/1182775715242967050/1193995186951426048/1235393057055182919))
>
> Also here are the [v. 8.1.240328-rc release notes](https://github.com/CommunityToolkit/Windows/releases/tag/v8.1.240328-rc)

**Reply by agneszitte** [reply]
@jeromelaban (cc @michael-hawker, @Arlodotexe) 
I think we can already update [our documentation](https://platform.uno/docs/articles/uno-community-toolkit.html?tabs=tabid-winui) without having to wait for WCT v8.1 stable as the WCT 8.0 packages and above already include direct support for Uno Platform. What do you think?
And to also update our related [WCT samples](https://github.com/unoplatform/Uno.Samples/blob/master/doc/samples.md#wct-datagrid) as needed, with a single project migration.

**Reply by tomchoondal** [reply]
@agneszitte @jeromelaban Finally I am able to use CommunityToolKit 7.1.2 into my Project, Thanks for the Help. 

But When I am trying to use the Shadow using below code

 "<Grid Grid.Row="1" >
   <Grid.Resources>
     <ui:AttachedDropShadow x:Key="CommonShadow" Offset="4" CastTo="{x:Bind ShadowTarget}" />
   </Grid.Resources>
   <Border x:Name="ShadowTarget" />
   <TextBlock ui:Effects.Shadow="{StaticResource CommonShadow}" Text="This is some text with a Shadow!" Foreground="Black"
       HorizontalAlignment="Center" />
 </Grid>"


![image](https://github.com/unoplatform/uno/assets/16680256/996c7dbb-c30a-4e09-8731-ad232a044025)


I am able to render in WinApp.  But in Web assembly it is not rendering and its throwing error in console 
![image](https://github.com/unoplatform/uno/assets/16680256/0894851c-8f21-4029-9345-172701f153cc)

Do you have any alternative for shadow in Community Toolkit?

**Reply by agneszitte** [reply]
Hi @tomchoondal, thank you for confirming that the setup for WCT is working with the details that I shared.
I have created issues so we can update the WCT documentation and related samples:
- https://github.com/unoplatform/uno/issues/16872
- https://github.com/unoplatform/Uno.Samples/issues/722

We already have an issue for what you encountered with the WCT `AttachedDropShadow`:
[Windows Community Toolkit AttachedCardShadow and AttachedDropShadow is not working](https://github.com/unoplatform/Uno.WindowsCommunityToolkit/issues/197#top)#197

Depending on what you are trying to achieve we have `ShadowContainer` that is available with the **Uno Toolkit** lib:
https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/controls/ShadowContainer.html
Here are the details to get started with **Uno Toolkit**:
https://platform.uno/docs/articles/external/uno.toolkit.ui/doc/getting-started.html

You can give it a try on our Gallery here: https://gallery.platform.uno/
![image](https://github.com/unoplatform/uno/assets/16295702/4fdef39c-e78d-4d92-b498-7a4976b95168)

**Reply by tomchoondal** [reply]
Tried above solution and its working fine. I am able to render shadow on WASM. Thank you for your support.

**Reply by agneszitte** [reply]
Glad to hear it @tomchoondal and thank you for using Uno Platform
